Unhandled exception when marker contains mesh from unknown package

When a marker msgs with a mesh from an unknown package is loaded, it throws an Ogre exception which is unfortunately not handled:
```
[ WARN] [1613474597.318764360]: OGRE EXCEPTION(6:FileNotFoundException): Cannot locate resource package://motek_demo_2017/meshes/tilted-base.stl in resource group Autodetect or any other group. in ResourceGroupManager::openResource at /build/o
gre-1.9-B6QkmW/ogre-1.9-1.9.0+dfsg1/OgreMain/src/OgreResourceGroupManager.cpp (line 756)                                                                                                                                                            
terminate called after throwing an instance of 'Ogre::FileNotFoundException'

  what():  OGRE EXCEPTION(6:FileNotFoundException): Cannot locate resource package://motek_demo_2017/meshes/tilted-base.stl in resource group Autodetect or any other group. in ResourceGroupManager::openResource at /build/ogre-1.9-B6QkmW/ogre-1
.9-1.9.0+dfsg1/OgreMain/src/OgreResourceGroupManager.cpp (line 756)


backtrace:


1   __GI_raise                                                                                                                                                                                                                                                                     raise.c 51 0x7f6ae9cadfb7 
2   __GI_abort                                                                                                                                                                                                                                                                     abort.c 79 0x7f6ae9caf921 
3   ??                                                                                                                                                                                                                                                                                        0x7f6aea6a2957 
4   ??                                                                                                                                                                                                                                                                                        0x7f6aea6a8ae6 
5   std::terminate()                                                                                                                                                                                                                                                                          0x7f6aea6a8b21 
6   __cxa_rethrow                                                                                                                                                                                                                                                                             0x7f6aea6a8da9 
7   Ogre::Resource::load(bool)                                                                                                                                                                                                                                                                0x7f6ae717fb08 
8   Ogre::MeshManager::load(std::string const&, std::string const&, Ogre::HardwareBuffer::Usage, Ogre::HardwareBuffer::Usage, bool, bool)                                                                                                                                                     0x7f6ae70ec0a7 
9   Ogre::EntityFactory::createInstanceImpl(std::string const&, std::map<std::string, std::string, std::less<std::string>, Ogre::STLAllocator<std::pair<std::string const, std::string>, Ogre::CategorisedAllocPolicy<(Ogre::MemoryCategory)0>>> const *)                                     0x7f6ae7013ce6 
10  Ogre::MovableObjectFactory::createInstance(std::string const&, Ogre::SceneManager *, std::map<std::string, std::string, std::less<std::string>, Ogre::STLAllocator<std::pair<std::string const, std::string>, Ogre::CategorisedAllocPolicy<(Ogre::MemoryCategory)0>>> const *)            0x7f6ae70fbd44 
11  Ogre::SceneManager::createMovableObject(std::string const&, std::string const&, std::map<std::string, std::string, std::less<std::string>, Ogre::STLAllocator<std::pair<std::string const, std::string>, Ogre::CategorisedAllocPolicy<(Ogre::MemoryCategory)0>>> const *)                 0x7f6ae71c477f 
12  Ogre::SceneManager::createEntity(std::string const&, std::string const&, std::string const&)                                                                                                                                                                                              0x7f6ae71c0f1a 
13  rviz::MeshResourceMarker::onNewMessage(boost::shared_ptr<visualization_msgs::Marker_<std::allocator<void>> const> const&, boost::shared_ptr<visualization_msgs::Marker_<std::allocator<void>> const> const&)                                                                              0x7f6adba6df9a 
14  rviz::MarkerBase::setMessage(boost::shared_ptr<visualization_msgs::Marker_<std::allocator<void>> const> const&)                                                                                                                                                                           0x7f6adba6b571 
15  rviz::MarkerDisplay::processAdd(boost::shared_ptr<visualization_msgs::Marker_<std::allocator<void>> const> const&)                                                                                                                                                                        0x7f6adba559e6 
16  rviz::MarkerDisplay::processMessage(boost::shared_ptr<visualization_msgs::Marker_<std::allocator<void>> const> const&)                                                                                                                                                                    0x7f6adba56103 
17  rviz::MarkerDisplay::update(float, float)                                                                                                                                                                                                                                                 0x7f6adba5651f 
18  rviz::DisplayGroup::update(float, float)                                                                                                                                                                                                                                                  0x7f6aed181b88 
19  rviz::VisualizationManager::onUpdate()                                                                                                                                                                                                                                                    0x7f6aed24bb25 
20  QMetaObject::activate(QObject *, int, int, void * *)                                                                                                                                                                                                                                      0x7f6af0631305 
... <Mehr>                                                                                                                                                                                                                                                                                                   



```

### Your environment
* OS Version: e.g. Ubuntu 18.04
* ROS Distro: [Kinetic | Melodic]
* RViz, Qt, OGRE, OpenGl version as printed by rviz:
    ```
    [ INFO] [1613475494.475253896]: rviz version 1.13.15
[ INFO] [1613475494.475289937]: compiled against Qt version 5.12.8
[ INFO] [1613475494.475299717]: compiled against OGRE version 1.9.0 (Ghadamon)
[ INFO] [1613475494.483962728]: Forcing OpenGl version 0.
[ INFO] [1613475495.046200133]: Stereo is NOT SUPPORTED
[ INFO] [1613475495.046340502]: OpenGl version: 3,0 (GLSL 1,3).

    ```
* If source build, which git commit? d4603a4f


I'll try to provide a PR soon

Strange one ... I also couldn't reproduce it in standalone rviz but it successfully crashes our UI that includes the `rviz::RenderPanel`

The crash also happens in `InteractiveMarker` 

comparing rviz and the embedded use case I see that rviz prints out the following for every marker, interactive marker, robot:
```
[rospack] Error: package 'motek_demo_2017' not found
[librospack]: error while executing command
[ERROR] [1615826641.080096501]: Could not load resource [package://motek_demo_2017/meshes/tilted-base.stl]: Unable to open file "package://motek_demo_2017/meshes/tilted-base.stl".

```

while rviz embedded in our application only prints that for `RobotStateDisplay` but not for the interactive marker where it crashes. So I wonder why this does not fail:
https://github.com/ros-visualization/rviz/blob/ba536b2d5baae4be167f9741a3f5bcb05d129af6/src/rviz/default_plugin/markers/mesh_resource_marker.cpp#L111

Here is another backtrace:
```
1  __GI_raise                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 raise.c      51  0x7f817c605fb7 
2  __GI_abort                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 abort.c      79  0x7f817c607921 
3  ??                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          0x7f817cffa957 
4  ??                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          0x7f817d000ae6 
5  std::terminate()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            0x7f817d000b21 
6  __cxa_rethrow                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               0x7f817d000da9 
7  Ogre::Resource::load(bool)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  0x7f8179ad7b08 
8  Ogre::MeshManager::load(std::string const&, std::string const&, Ogre::HardwareBuffer::Usage, Ogre::HardwareBuffer::Usage, bool, bool)                                                                                                                                                                                                                                                                                                                                                                                                                                                                       0x7f8179a440a7 
9  Ogre::EntityFactory::createInstanceImpl(std::string const&, std::map<std::string, std::string, std::less<std::string>, Ogre::STLAllocator<std::pair<std::string const, std::string>, Ogre::CategorisedAllocPolicy<(Ogre::MemoryCategory)0>>> const *)                                                                                                                                                                                                                                                                                                                                                       0x7f817996bce6 
10 Ogre::MovableObjectFactory::createInstance(std::string const&, Ogre::SceneManager *, std::map<std::string, std::string, std::less<std::string>, Ogre::STLAllocator<std::pair<std::string const, std::string>, Ogre::CategorisedAllocPolicy<(Ogre::MemoryCategory)0>>> const *)                                                                                                                                                                                                                                                                                                                              0x7f8179a53d44 
11 Ogre::SceneManager::createMovableObject(std::string const&, std::string const&, std::map<std::string, std::string, std::less<std::string>, Ogre::STLAllocator<std::pair<std::string const, std::string>, Ogre::CategorisedAllocPolicy<(Ogre::MemoryCategory)0>>> const *)                                                                                                                                                                                                                                                                                                                                   0x7f8179b1c77f 
12 Ogre::SceneManager::createEntity(std::string const&, std::string const&, std::string const&)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                0x7f8179b18f1a 
13 rviz::MeshResourceMarker::onNewMessage(boost::shared_ptr<visualization_msgs::Marker_<std::allocator<void>> const> const&, boost::shared_ptr<visualization_msgs::Marker_<std::allocator<void>> const> const&)                                                                                                                                                                                                                                                                                                                                                                                                0x7f816e3c5f9a 
14 rviz::MarkerBase::setMessage(boost::shared_ptr<visualization_msgs::Marker_<std::allocator<void>> const> const&)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             0x7f816e3c3571 
15 rviz::InteractiveMarkerControl::makeMarkers(visualization_msgs::InteractiveMarkerControl_<std::allocator<void>> const&)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     0x7f816e383be7 
16 rviz::InteractiveMarkerControl::processMessage(visualization_msgs::InteractiveMarkerControl_<std::allocator<void>> const&)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  0x7f816e3844a0 
17 rviz::InteractiveMarker::processMessage(visualization_msgs::InteractiveMarker_<std::allocator<void>> const&)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                0x7f816e38a9af 
18 rviz::InteractiveMarkerDisplay::updateMarkers(std::string const&, std::vector<visualization_msgs::InteractiveMarker_<std::allocator<void>>, std::allocator<visualization_msgs::InteractiveMarker_<std::allocator<void>>>> const&)                                                                                                                                                                                                                                                                                                                                                                           0x7f816e37a8cf 
19 rviz::InteractiveMarkerDisplay::updateCb(boost::shared_ptr<visualization_msgs::InteractiveMarkerUpdate_<std::allocator<void>> const>)                                                                                                                                                                                                                                                                                                                                                                                                                                                                       0x7f816e37b9b8 
20 boost::detail::function::void_function_obj_invoker1<boost::_bi::bind_t<void, boost::_mfi::mf1<void, rviz::InteractiveMarkerDisplay, boost::shared_ptr<visualization_msgs::InteractiveMarkerUpdate_<std::allocator<void>> const>>, boost::_bi::list2<boost::_bi::value<rviz::InteractiveMarkerDisplay *>, boost::arg<1>>>, void, boost::shared_ptr<visualization_msgs::InteractiveMarkerUpdate_<std::allocator<void>> const> const&>::invoke(boost::detail::function::function_buffer&, boost::shared_ptr<visualization_msgs::InteractiveMarkerUpdate_<std::allocator<void>> const> const&)                  0x7f816e37c1ca 
21 interactive_markers::SingleClient::pushUpdates()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            0x7f817f54ed3e 
22 interactive_markers::SingleClient::update()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 0x7f817f5503a5 
23 interactive_markers::InteractiveMarkerClient::update()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      0x7f817f545359 
24 rviz::InteractiveMarkerDisplay::update(float, float)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        0x7f816e377626 
25 rviz::DisplayGroup::update(float, float)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    0x7f817fad9b88 
26 rviz::VisualizationManager::onUpdate()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      0x7f817fba3b25 
27 QMetaObject::activate(QObject *, int, int, void * *)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        0x7f8182f89305 
28 QTimer::timeout(QTimer::QPrivateSignal)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     0x7f8182f96237 
29 QTimer::timerEvent(QTimerEvent *)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           0x7f8182f96598 
30 QObject::event(QEvent *)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    0x7f8182f89e6b 
31 QApplicationPrivate::notify_helper(QObject *, QEvent *)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     0x7f81821d8e4c 
32 QApplication::notify(QObject *, QEvent *)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   0x7f81821e0450 
33 QCoreApplication::notifyInternal2(QObject *, QEvent *)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      0x7f8182f597d8 
34 QTimerInfoList::activateTimers()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            0x7f8182fb5cf9 
35 ??                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          0x7f8182fb64c1 
36 g_main_context_dispatch                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     0x7f817ab53537 
37 ??                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          0x7f817ab53770 
38 g_main_context_iteration                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    0x7f817ab537fc 
39 QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  0x7f8182fb688f 
40 QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     0x7f8182f57a2a 
41 QCoreApplication::exec()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    0x7f8182f60c30 
42 ??                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          0x561bb3411d23 
43 __libc_start_main                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          libc-start.c 310 0x7f817c5e8bf7 
44 ??                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          0x561bb342435a 
```

I'm not sure how to debug this further and have some headache due to covid so I'd be grateful for ideas. I noticed that `InteractiveMarker` just silently ignores (from a GUI point of view) the broken markers but still it does not crash in rviz. I could not find a general purpose `catch` that would explain it.

--- a/src/rviz/default_plugin/robot_model_display.cpp
+++ b/src/rviz/default_plugin/robot_model_display.cpp
@@ -35,6 +35,7 @@
 
 #include "rviz/display_context.h"
 #include "rviz/robot/robot.h"
+#include "rviz/robot/robot_link.h"
 #include "rviz/robot/tf_link_updater.h"
 #include "rviz/properties/float_property.h"
 #include "rviz/properties/property.h"
@@ -192,6 +193,17 @@ void RobotModelDisplay::load()
 
   setStatus(StatusProperty::Ok, "URDF", "URDF parsed OK");
   robot_->load(descr);
+  std::stringstream ss;
+  for (const auto& name_link_pair : robot_->getLinks())
+  {
+    const std::string& err = name_link_pair.second->getGeometryErrors();
+    if (!err.empty())
+      ss << "\n• for link '" << name_link_pair.first << "':\n" << err;
+  }
+  if (ss.tellp())
+    setStatus(StatusProperty::Error, "URDF",
+              QString("Errors loading geometries:").append(ss.str().c_str()));
+
   robot_->update(TFLinkUpdater(context_->getFrameManager(),
                                boost::bind(linkUpdaterStatusFunction, _1, _2, _3, this),
                                tf_prefix_property_->getStdString()));

--- a/src/rviz/properties/property_tree_with_help.cpp
+++ b/src/rviz/properties/property_tree_with_help.cpp
@@ -66,6 +66,7 @@ void PropertyTreeWithHelp::showHelpForProperty(const Property* property)
   {
     QString body_text = property->getDescription();
     QString heading = property->getName();
+    body_text.replace("\n", "<br>");
     QString html = "<html><body><strong>" + heading + "</strong><br>" + body_text + "</body></html>";
     help_->setHtml(html);
   }

--- a/src/rviz/robot/robot_link.cpp
+++ b/src/rviz/robot/robot_link.cpp
@@ -151,6 +151,7 @@ void RobotLinkSelectionHandler::postRenderPass(uint32_t /*pass*/)
   }
 }
 
+static std::map<const RobotLink*, std::string> errors;
 
 RobotLink::RobotLink(Robot* robot,
                      const urdf::LinkConstSharedPtr& link,
@@ -318,6 +319,26 @@ RobotLink::~RobotLink()
   delete axes_;
   delete details_;
   delete link_property_;
+  errors.erase(this);
+}
+
+void RobotLink::addError(const char* format, ...)
+{
+  char buffer[256];
+  va_list args;
+  va_start(args, format);
+  vsnprintf(buffer, sizeof(buffer), format, args);
+  va_end(args);
+
+  std::string& err = const_cast<std::string&>(getGeometryErrors());
+  if (!err.empty())
+    err.append("\n");
+  err.append(buffer);
+}
+
+const std::string& RobotLink::getGeometryErrors() const
+{
+  return errors[this];
 }
 
 bool RobotLink::hasGeometry() const
@@ -595,25 +616,22 @@ void RobotLink::createEntityForGeometryElement(const urdf::LinkConstSharedPtr& l
 
     scale = Ogre::Vector3(mesh.scale.x, mesh.scale.y, mesh.scale.z);
 
-    std::string model_name = mesh.filename;
+    const std::string& model_name = mesh.filename;
 
     try
     {
       if (loadMeshFromResource(model_name).isNull())
-        ROS_ERROR("Failed to load mesh resource '%s' for link '%s'.", model_name.c_str(),
-                  link->name.c_str());
+        addError("Could not load mesh resource '%s'", model_name.c_str());
       else
         entity = scene_manager_->createEntity(ss.str(), model_name);
     }
     catch (Ogre::InvalidParametersException& e)
     {
-      ROS_ERROR("Could not convert mesh resource '%s' for link '%s'. It might be an empty mesh: %s",
-                model_name.c_str(), link->name.c_str(), e.what());
+      addError("Could not convert mesh resource '%s': %s", model_name.c_str(), e.what());
     }
     catch (Ogre::Exception& e)
     {
-      ROS_ERROR("Could not load model '%s' for link '%s': %s", model_name.c_str(), link->name.c_str(),
-                e.what());
+      addError("Could not load model '%s': %s", model_name.c_str(), e.what());
     }
     break;
   }

--- a/src/rviz/robot/robot_link.cpp
+++ b/src/rviz/robot/robot_link.cpp
@@ -716,8 +716,8 @@ void RobotLink::createCollision(const urdf::LinkConstSharedPtr& link)
       if (collision_mesh)
       {
         collision_meshes_.push_back(collision_mesh);
-        valid_collision_found = true;
       }
+      valid_collision_found |= collision == link->collision; // don't consider the same geometry twice
     }
   }
 #endif
@@ -776,8 +776,8 @@ void RobotLink::createVisual(const urdf::LinkConstSharedPtr& link)
       if (visual_mesh)
       {
         visual_meshes_.push_back(visual_mesh);
-        valid_visual_found = true;
       }
+      valid_visual_found |= visual == link->visual; // don't consider the same geometry again
     }
   }
 #endif

--- a/src/rviz/robot/robot_link.cpp
+++ b/src/rviz/robot/robot_link.cpp
@@ -599,8 +599,11 @@ void RobotLink::createEntityForGeometryElement(const urdf::LinkConstSharedPtr& l
 
     try
     {
-      loadMeshFromResource(model_name);
-      entity = scene_manager_->createEntity(ss.str(), model_name);
+      if (loadMeshFromResource(model_name).isNull())
+        ROS_ERROR("Failed to load mesh resource '%s' for link '%s'.", model_name.c_str(),
+                  link->name.c_str());
+      else
+        entity = scene_manager_->createEntity(ss.str(), model_name);
     }
     catch (Ogre::InvalidParametersException& e)
     {

--- a/src/rviz/robot/robot_link.h
+++ b/src/rviz/robot/robot_link.h
@@ -128,6 +128,7 @@ public:
   {
     return robot_;
   }
+  const std::string& getGeometryErrors() const;
 
   // Remove link_property_ from its old parent and add to new_parent.  If new_parent==NULL then leav
   // unparented.
@@ -189,6 +190,7 @@ private:
                                       const urdf::Pose& origin,
                                       Ogre::SceneNode* scene_node,
                                       Ogre::Entity*& entity);
+  void addError(const char* format, ...);
 
   void createVisual(const urdf::LinkConstSharedPtr& link);
   void createCollision(const urdf::LinkConstSharedPtr& link);

